Math/NumberTheory/SmoothNumbers.hs

```diff
@@ -76,29 +80,42 @@ fromSmoothUpperBound n = if (n < 2)
                          then Nothing
                          else Just $ SmoothBasis $ takeWhile (<= n) $ map unPrime primes
 
--- | Generate an infinite ascending list of
--- <https://en.wikipedia.org/wiki/Smooth_number smooth numbers>
--- over a given smooth basis.
---
--- >>> import Data.Maybe
--- >>> take 10 (smoothOver (fromJust (fromList [2, 5])))
--- [1, 2, 4, 5, 8, 10, 16, 20, 25, 32]
-smoothOver :: Integral a => SmoothBasis a -> [a]
-smoothOver pl = foldr (\p l -> mergeListLists $ iterate (map (p*)) l) [1] (unSmoothBasis pl)
+-- | Helper used by @smoothOver@ (@Integral@ constraint) and @smoothOver'@
+-- (@Euclidean@ constraint) Since the typeclass constraint is just
+-- @Num@, it receives a @norm@ comparison function for the generated smooth
+-- numbers.
+-- This function relies on the fact that for any element of a smooth basis @p@
+-- and any @a@ it is true that @norm (a * p) > norm a@.
+-- This condition is not checked.
+smoothOver'' :: forall a b . (Num a, Ord b) => (a -> b) -> SmoothBasis a -> [a]
```

Comment:
Wow! This is existing. Actually, I did not realise that `Euclidean` constraint is redundant here and `Num` is sufficient. Let us stick to `Num` in other function as well.

Reply:
>existing

What do you mean here?

Reply:
Sorry, just a silly typo.

Reply:
@Bodigrim won't dropping the `Integral` constraint in `smoothOver` mean `abs` can no longer be used as norm?

Reply:
Let us leave `smoothOver` with `Integral`, but make `smoothOver' == smoothOver''`.